File: python/src/day04.py
#!/usr/bin/env python

import re
from pathlib import Path
from typing import Dict, List

Passport = Dict[str, str]

INPUT_FILE = Path.cwd().parent / "inputs" / "day04.txt"
ECL_COLORS = {"amb", "blu", "brn", "gry", "grn", "hzl", "oth"}

VALIDATORS = {
    "byr": lambda x: 1920 <= int(x) <= 2002,
    "iyr": lambda x: 2010 <= int(x) <= 2020,
    "eyr": lambda x: 2020 <= int(x) <= 2030,
    "hgt": lambda x: validate_hgt(x),
    "hcl": lambda x: re.match(r"^#[0-9a-f]{6}$", x),
    "ecl": lambda x: x in ECL_COLORS,
    "pid": lambda x: re.match(r"^\d{9}$", x),
}


def validate_hgt(x: str) -> bool:
    if x[-2:] == "cm":
        return 150 <= int(x[:-2]) <= 193
    if x[-2:] == "in":
        return 59 <= int(x[:-2]) <= 76
    return False


def transform_input(input: str) -> List[Passport]:
    passports = []

    for line in input.split("\n\n"):
        passport = {}
        for word in line.split():
            try:
                key, val = word.split(":", 1)
                passport[key] = val
            except ValueError:
                pass
        passports.append(passport)

    return passports


def part1(passports: List[Passport]) -> int:
    return sum(check_passport_keys(x) for x in passports)


def check_passport_keys(passport: Passport) -> bool:
    return all(key in passport for key in VALIDATORS.keys())


def part2(passports: List[Passport]) -> int:
    return sum(check_passport_full(x) for x in passports)


def check_passport_full(passport: Passport) -> bool:
    for key, validator in VALIDATORS.items():
        try:
            if not validator(passport[key]):
                return False
        except (ValueError, KeyError):
            return False

    return True


if __name__ == "__main__":
    with INPUT_FILE.open("r") as f:
        input = transform_input(f.read())

    print(f"Part 1: {part1(input)}")
    print(f"Part 2: {part2(input)}")

File: rust/src/day04.rs
use itertools::Itertools;
use lazy_static::lazy_static;
use regex::Regex;
use std::{collections::HashMap, str::FromStr};

static INPUT: &str = include_str!("../../inputs/day04.txt");
static ECL_COLORS: [&str; 7] = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"];

lazy_static! {
    static ref HCL_REGEX: Regex = Regex::new(r"^#[0-9a-f]{6}$").unwrap();
    static ref PID_REGEX: Regex = Regex::new(r"^\d{9}$").unwrap();
}

type Passport<'a> = HashMap<&'a str, &'a str>;
type ValidatorResult = Result<bool, <i32 as FromStr>::Err>;
type Validator<'a> = (&'a str, Box<dyn Fn(&'a str) -> ValidatorResult>);

pub fn run() {
    let passports = transform_input(INPUT);

    let validators: Vec<Validator> = vec![
        ("byr", Box::new(|x| check_range(x, 1920, 2002))),
        ("iyr", Box::new(|x| check_range(x, 2010, 2020))),
        ("eyr", Box::new(|x| check_range(x, 2020, 2030))),
        (
            "hgt",
            Box::new(|x| {
                if x.ends_with("cm") {
                    check_range(x.get(0..x.len() - 2).unwrap(), 150, 193)
                } else if x.ends_with("in") {
                    check_range(x.get(0..x.len() - 2).unwrap(), 59, 76)
                } else {
                    Ok(false)
                }
            }),
        ),
        ("hcl", Box::new(|x| Ok(HCL_REGEX.is_match(x)))),
        ("ecl", Box::new(|x| Ok(ECL_COLORS.contains(&x)))),
        ("pid", Box::new(|x| Ok(PID_REGEX.is_match(x)))),
    ];

    println!("Part 1: {}", part1(&passports, &validators));
    println!("Part 2: {}", part2(&passports, &validators));
}

fn check_range(value: &str, lower: i32, upper: i32) -> ValidatorResult {
    let value: i32 = value.parse()?;
    Ok(value >= lower && value <= upper)
}

fn transform_input(input: &str) -> Vec<Passport> {
    input
        .split("\n\n")
        .map(|passport| {
            passport
                .split_whitespace()
                .map(|x| x.splitn(2, ':').collect_tuple().unwrap())
                .collect()
        })
        .collect()
}

fn part1(passports: &[Passport], validators: &[Validator]) -> usize {
    passports
        .iter()
        .filter(|passport| validators.iter().all(|(key, _)| passport.contains_key(key)))
        .count()
}

fn part2<'a>(passports: &'a [Passport], validators: &[Validator<'a>]) -> usize {
    passports
        .iter()
        .filter(|&passport| {
            validators.iter().all(|(key, validator)| {
                passport.contains_key(key)
                    && Ok(true) == validator(&passport.get(key).unwrap())
            })
        })
        .count()
}
